dimod/include/dimod/abc.h

```diff
@@ -0,0 +1,1083 @@
+            this->linear_biases_ = std::move(other.linear_biases_);
+            this->adj_ptr_ = std::move(other.adj_ptr_);
+            this->offset_ = other.offset_;
+        }
+        return *this;
+    }
+
+    explicit QuadraticModelBase(std::vector<bias_type>&& linear_biases)
+            : linear_biases_(linear_biases), adj_ptr_(), offset_(0) {}
+
+    explicit QuadraticModelBase(index_type n) : linear_biases_(n), adj_ptr_(), offset_(0) {}
+
+    /// Increase the size of the model by one. Returns the index of the new variable.
+    index_type add_variable();
+
+    /// Increase the size of the model by `n`. Returns the index of the first variable added.
+    index_type add_variables(index_type n);
+
+    /**
+     * Adjust the biases of a single variable to reinterpret it as being a
+     * different vartype.
+     *
+     * This method is meant to be used by subclasses, since it does not check
+     * or update the actual vartypes of the variables.
+     */
+    void change_vartype(Vartype source, Vartype target, index_type v);
+
+    /**
+     * Adjust all of the biases in the model to reinterpret them as being
+     * from the source vartype to the target.
+     *
+     * This method is meant to be used by subclasses, since it does not check
+     * or update the actual vartypes of the variables.
+     */
+    void change_vartypes(Vartype source, Vartype target);
+
+    /// Resize model to contain n variables.
+    void resize(index_type n);
+
+ private:
+    std::vector<bias_type> linear_biases_;
+
+    std::unique_ptr<std::vector<std::vector<LinearTerm<bias_type, index_type>>>> adj_ptr_;
+
+    bias_type offset_;
+
+    // Assumes adj exists!
+    // Creates the bias if it doesn't already exist
+    bias_type& asym_quadratic(index_type u, index_type v) {
+        assert(0 <= u && static_cast<size_type>(u) < num_variables());
+        assert(0 <= v && static_cast<size_type>(v) < num_variables());
+        assert(has_adj());
+
+        auto& neighborhood = (*adj_ptr_)[u];
+        auto it = std::lower_bound(neighborhood.begin(), neighborhood.end(), v);
+        if (it == neighborhood.end() || it->v != v) {
+            // we could make a bunch individual functions to avoid needing to
+            // default to 0, but this is a lot simpler.
+            it = neighborhood.emplace(it, v, 0);
+        }
+        return it->bias;
+    }
+
+    /// Create the adjacency structure if it doesn't already exist.
+    void enforce_adj() {
+        if (!adj_ptr_) {
+            adj_ptr_ = std::unique_ptr<std::vector<std::vector<LinearTerm<bias_type, index_type>>>>(
+                    new std::vector<std::vector<LinearTerm<bias_type, index_type>>>(
+                            num_variables()));
+        }
+    }
+
+    /// Return an empty neighborhood - useful when we don't have an adj
+    static const std::vector<LinearTerm<bias_type, index_type>>& empty_neighborhood() {
+        static std::vector<LinearTerm<bias_type, index_type>> empty;
```

Comment:
static ? Is it intenional ?

Reply:
Yes, we only want one to be allocated, and then we can refer to it globally.